### ISSUE TITLE
[Proposal]: Dapr as a capability delivery vehicle

### ISSUE THREAD

**Issue by Haishi2016** [dapr/dapr#1311]
# Overview
Dapr is designed to be a sidecar that delivers distributed system capabilities, such as pub-sub and state management, to applications. Currently, the set of capabilities are collectively decided by the Dapr community and baked into the Dapr runtime. This proposal extends the scope of capability delivery to allow new capabilities to be described and delivered as add-on components to Dapr without modifying the Dapr core runtime. For instance, if an application needs object detection capability, it can leverage a "capability package"  that Dapr will ensure available to be consumed. 

## Capability package
A capability package is described by a capability manifest, which defines:
* The package binary. As a starting point this is likely to be a Docker container.
* The API definition, such as a Swagger spec. This is mostly to help developer to understand how the API is to be used.
* Additional configurations such as authentication.

## Dapr initialization
When Dapr is initialized, it checks configured capability packages and creates necessary deployments to launch the corresponding containers. Then, it configure its routes to include something like \<version\>/capability/\<package\>/api. Then, all requests to the route will be forwarded to corresponding capability package.

## Benefit
* Allow capabilities to be defined and plugged into Dapr at runtime. 
* Dapr provides mTLS, distributed tracing, retries, rate limiting etc. out-of-box.
* When multiple vendors chose to implement the same API surface, it's easy to switch to different implementations through Dapr.

## Future
The capability package can be extended to include multiple service endpoints / packages in a priority list. When services with higher priority are unavailable, Dapr can fall through the chain to deliver the required capability. This is a very powerful feature for edge scenarios. For example, in a occasionally connected edge scenario, Dapr can provide business continuity by falling back from a cloud-hosted service to a local service when connection is broken.
The prioritization can be done in different ways, such by SLA promises, latency, cost and vendor. And policies can be reinforced to ensure only service providers that satisfy the non-functional requirements are used.

## Future Future
There are some thoughts on making capability automatically discoverable and consumable but we can leave that for the future.

**Comment by krautbax** [comment]
So, for N capabilities, there could be in theory an additional N Docker containers running in a k8s pod, for example, along with the application and daprd?  Am I understanding correctly?

**Comment by krautbax** [comment]
Assuming we would address binary compatibility, what about an in-process capability option using plugins (https://golang.org/pkg/plugin/)?

**Comment by Haishi2016** [comment]
Indeed. Ideally once plugins are supported on Windows.

**Comment by krautbax** [comment]
Fair point @Haishi2016 ("Currently plugins are only supported on Linux, FreeBSD, and macOS").
